File: binary_search_tree.py
#!/usr/bin/env python
# -*- coding=utf-8 -*-

class TreeNode:
    def __init__(self, elem, left=None, right=None):
        self.data = elem
        self.left = left
        self.right = right

def add(root, elem):
    if not root:
        return TreeNode(elem)

    if elem < root.data:
        node = add(root.left, elem)
        root.left = node
    elif elem > root.data:
        node = add(root.right, elem)
        root.right = node
    return root

def search(root, elem):
    if not root:
        return None

    if root.data == elem:
        return root
    elif elem < root.data:
        return search(root.left, elem)
    elif elem > root.data:
        return search(root.right, elem)

def inorder_print(root):
    if root:
        inorder_print(root.left)
        print(root.data, end=' ')
        inorder_print(root.right)

class BinarySearchTree:
    def __init__(self):
        self.num_nodes = 0
        self.root = None

    def is_empty(self):
        return self.size() == 0

    def size(self):
        return self.num_nodes

    def add(self, elem):
        if self.contains(elem):
            return False
        self.root = self._add(self.root, elem)
        self.num_nodes += 1
        return True

    # Recursively add a value in the binary tree
    def _add(self, node, elem):
        # Base case: we found a leaf node
        if node is None:
            node = TreeNode(elem)
        else:
            if elem < node.data:
                node.left = self._add(node.left, elem)
            else:
                node.right = self._add(node.right, elem)
        return node

    # Return the node if the element exists in the tree
    def search(self, elem):
        return self._search(self.root, elem)

    def _search(self, node, elem):
        if node is None:
            return None
        elif elem == node.data:
            return node
        elif elem < node.data:
            return self._search(node.left, elem)
        else:
            return self._search(node.right, elem)

    # Returns True if the element exists in the tree
    def contains(self, elem):
        return self._contains(self.root, elem)

    def _contains(self, node, elem):
        if node is None:
            return False
        
        if elem < node.data:
            return self._contains(node.left, elem)
        elif elem > node.data:
            return self._contains(node.right, elem)
        else:
            return True

    def remove(self, elem):
        if self.contains(elem):
            self.root = self._remove(self.root, elem)
            self.num_nodes -= 1
            return True
        else:
            return False

    def _remove(self, node, elem):
        if node is None:
            return None

        # Dig into left subtree, the value we're looking for is
        # samller than the current value
        if node.data > elem:
            node.left = self._remove(node.left, elem)
        # Dig into right subtree, the value we're looking for is
        # larger than the current value
        elif node.data < elem:
            node.right = self._remove(node.right, elem)
        # Found the node we wish to remove
        else:
            # In this situation just swap the node we wish to remove
            # with its right child
            # if right child is also None, return None(i.e. remove leaf node)
            if node.left is None:
                return node.right
            # In this situation just swap the node we wish to remove
            # with its left child
            elif node.right is None:
                return node.left
            # when removing a node from a binary tree with two links
            # the successor of the node being removed can either be the
            # largest value in the left subtree or the smallest value in the right subtree.
            # In this implementation, I have decided to find the smallest value in the right subtree
            # which can be found by traversing as far as possible in the right subtree.
            else:
                # Find the leftmost node in the right subtree
                tmp = self.dig_left(node.right)

                # swap the data
                node.data = tmp

                # Go into the right subtree and remove the leftmost
                # node we found and swapped data with.
                # This prevents us from having two nodes in our tree with the same value
                node.right = self._remove(node.right, tmp.data)

        return node

    def dig_left(self, node):
        cur = node
        while cur.left:
            cur = cur.left
        return cur

    def inorder_print(self):
        self._inorder_print(self.root)
        print('')

    def _inorder_print(self, node):
        if node:
            self._inorder_print(node.left)
            print(node.data, end=' ')
            self._inorder_print(node.right)

    def levelorder_print(self):
        q = [self.root]
        while q:
            node = q.pop(0)
            print(node.data, end=' ')
            q = q + [node.left] if node.left else q
            q = q + [node.right] if node.right else q
        print('')

if __name__ == "__main__":
    bst = BinarySearchTree()
    print('is_empty :', bst.is_empty())
    bst.add(10)
    bst.add(4)
    bst.add(1)
    bst.add(5)
    bst.add(6)
    bst.add(20)
    bst.add(15)
    bst.add(33)
    print('inorder_print:')
    bst.inorder_print()
    print('levelorder_print:')
    bst.levelorder_print()

    target = 10
    if bst.search(target):
        print('Search {} -> Found'.format(target))
    else:
        print('Search {} -> Not found'.format(target))

    target = -100
    if bst.search(target):
        print('Search {} -> Found'.format(target))
    else:
        print('Search {} -> Not found'.format(target))

    print('remove(1):')
    if bst.remove(1):
        bst.levelorder_print()
    else:
        print('remove failed.')

    root = None
    root = add(root, 3)
    inorder_print(root)
    print()
    root = add(root, 2)
    inorder_print(root)
    print()
    root = add(root, 5)
    inorder_print(root)
    print()
    root = add(root, -1)
    inorder_print(root)
    print()
    root = add(root, 0)
    inorder_print(root)
    print()
    root = add(root, 10)
    inorder_print(root)
    print()

    target = -1
    node = search(root, target)
    if node:
        print('Found: {}'.format(node.data))
    else:
        print('Not found(search {})'.format(target))

File: binary_search_tree.cpp
#include <iostream>

class TreeNode {
    public:
        int data;
        TreeNode* left;
        TreeNode* right;

    TreeNode() {
        data = 0;
        left = nullptr;
        right = nullptr;
    }

    TreeNode(int elem) {
        data = elem;
        left = nullptr;
        right = nullptr;
    }
};

class BinarySearchTree {
    TreeNode* root;
    int num_nodes;

    public:
        BinarySearchTree(int elem) {
            root = new TreeNode(elem);
            num_nodes = 1;
        }

        TreeNode* getRoot() {
            return root;
        }

        bool remove_leaf(TreeNode* parent, TreeNode* node) {
            // Delete the leaf node by making the leaf node's parent's left or right child is nullptr
            if (root->data == node->data) {
                delete root;
                root = nullptr;
                return true;
            } else if (node->data < parent->data) {
                delete parent->left;
                parent->left = nullptr;
                return true;
            } else {
                delete parent->right;
                parent->right = nullptr;
                return true;
            }
        }

        bool remove_node_with_left_only(TreeNode* parent, TreeNode* node) {
            // Delete the node by making the left child of that node the node's parent's right or left child
            if (root->data == node->data) {
                delete root;
                root = node->left;
                return true;
            } else if (node->data < parent->data) {
                delete parent->left;
                parent->left = node->left;
                return true;
            } else {
                delete parent->right;
                parent->right = node->left;
                return true;
            }
        }

        bool remove_node_with_right_only(TreeNode* parent, TreeNode* node) {
            // Similar to the above left only case
            if (root->data == node->data) {
                delete root;
                root = node->right;
                return true;
            } else if (node->data < parent->data) {
                delete parent->left;
                parent->left = node->right;
                return true;
            } else {
                delete parent->right;
                parent->right = node->right;
                return true;
            }
        }

        bool remove_node_with_two_children(TreeNode* parent, TreeNode* node) {
            // Find the smallest value node in the right subtree of the current node
            TreeNode* least_node = findLeastNode(node);
            int data = least_node->data;
            // Swap the current value with the samllest value
            bool is_deleted = remove(root, data);
            node->data = data;
            return is_deleted;
        }

        TreeNode* findLeastNode(TreeNode* node) {
            TreeNode* current = node;
            while (current->left != nullptr) {
                current = current->left;
            }
            return current;
        }

        bool remove(TreeNode* node, int elem) {
            if (root == nullptr) {
                return false;
            }

            // Search the elem
            TreeNode* parent;
            while (node != nullptr && node->data != elem) {
                parent = node;
                if (elem < node->data) {
                    node = node->left;
                } else {
                    node = node->right;
                }
            }

            if (node == nullptr) {
                // Return false when the elem is not found
                return false;
            } else if (node->left == nullptr and node->right == nullptr) {
                // Deleting a leaf node
                return remove_leaf(parent, node);
            } else if (node->right == nullptr) {
                // Deleting a node with a left child only
                return remove_node_with_left_only(parent, node);
            } else if (node->left == nullptr) {
                // Deleting a node with a right child only
                return remove_node_with_right_only(parent, node);
            } else {
                // Deleting a node with two children
                return remove_node_with_two_children(parent, node);
            }
        }

        TreeNode* search(int elem) {
            TreeNode* current_node = root;
            while (current_node && current_node->data != elem) {
                if (elem < current_node->data) {
                    current_node = current_node->left;
                } else {
                    current_node = current_node->right;
                }
            }
            // after the loop, we'll have either the searched value
            // or nullptr in case the value was not found
            return current_node;
        }

        TreeNode* search(TreeNode* node, int elem) {
            if (node == nullptr) {
                return nullptr;
            } else if (node->data == elem) {
                return node;
            } else if (elem < node->data) {
                return search(node->left, elem);
            } else {
                return search(node->right, elem);
            }
        }

        bool contains(int elem) {
            return contains_r(root, elem);
        }

        bool contains_r(TreeNode* node, int elem) {
            if (node == nullptr) {
                return false;
            }

            if (elem < node->data) {
                return contains_r(node->left, elem);
            } else if (elem > node->data)
            {
                return contains_r(node->right, elem);
            } else {
                return true;
            }
        }

        bool addNode(int elem) {
            if (contains(elem)) {
                return false;
            } else {
                root = addNode_r(root, elem);
                num_nodes++;
                return true;
            }
        }

        TreeNode* addNode_r(TreeNode* node, int elem) {
            // base case
            // create new node and return it to the previous leaf(null) node.
            if (node == nullptr) {
                node = new TreeNode(elem);
            } else if (elem < node->data) {
                node->left = addNode_r(node->left, elem);
            } else {
                node->right = addNode_r(node->right, elem);
            }
            return node;
        }

        bool addNode_i(int elem) {
            if (contains(elem)) {
                return false;
            }

            if (getRoot() == nullptr) {
                root = new TreeNode(elem);
                num_nodes++;
                return true;
            }

            // Starting from the root
            TreeNode* node = root;
            TreeNode* parent;
            while (node) {
                parent = node;
                if (elem < node->data) {
                    node = node->left;
                } else {
                    node = node->right;
                }
            }
            // When the current node becomes nullptr, it inserts the value to its parent.
            if (elem < parent->data) {
                parent->left = new TreeNode(elem);
            } else {
                parent->right = new TreeNode(elem);
            }

            num_nodes++;
            return true;
        }

        void inOrderPrint(TreeNode* node) {
            if (node != nullptr) {
                inOrderPrint(node->left);
                std::cout << node->data << std::endl;
                inOrderPrint(node->right);
            }
        }

        void preOrderPrint(TreeNode* node) {
            if (node != nullptr) {
                std::cout << node->data << std::endl;
                preOrderPrint(node->left);
                preOrderPrint(node->right);
            }
        }

        void postOrderPrint(TreeNode* node) {
            if (node != nullptr) {
                postOrderPrint(node->left);
                postOrderPrint(node->right);
                std::cout << node->data << std::endl;
            }
        }
};

int main() {
    BinarySearchTree bst(6);
    std::cout << "root value : " << bst.getRoot()->data << std::endl;
    bst.addNode(4);
    bst.addNode(9);
    bst.addNode(5);
    bst.addNode(2);
    bst.addNode(8);
    bst.addNode(12);
    std::cout << "inorder print" << std::endl;
    bst.inOrderPrint(bst.getRoot());
    std::cout << "preorder print" << std::endl;
    bst.preOrderPrint(bst.getRoot());
    std::cout << "postorder print" << std::endl;
    bst.postOrderPrint(bst.getRoot());

    BinarySearchTree bst2(6);
    std::cout << "root value : " << bst2.getRoot()->data << std::endl;
    bst2.addNode(4);
    bst2.addNode(9);
    bst2.addNode(5);
    bst2.addNode(2);
    bst2.addNode(8);
    bst2.addNode(12);
    std::cout << "inorder print" << std::endl;
    bst2.inOrderPrint(bst2.getRoot());
    std::cout << "preorder print" << std::endl;
    bst2.preOrderPrint(bst2.getRoot());
    std::cout << "post order print" << std::endl;
    bst2.postOrderPrint(bst2.getRoot());

    int target = 12;
    if (bst.search(target)) {
        printf("Recursive search %d -> Found\n", target);
    } else {
        printf("Recursive search %d -> Not found\n", target);
    }

    if (bst.search(bst.getRoot(), target)) {
        printf("Iterative search %d -> Found\n", target);
    } else {
        printf("Iterative search %d -> Not found\n", target);
    }

    int delete_target = 5;
    if (bst.remove(bst.getRoot(), delete_target)) {
        printf("Delete %d\n", delete_target);
        bst.inOrderPrint(bst.getRoot());
    } else {
        printf("Not found %d\n", delete_target);
    }

    delete_target = -1;
    if (bst.remove(bst.getRoot(), delete_target)) {
        printf("Delete %d\n", delete_target);
        bst.inOrderPrint(bst.getRoot());
    } else {
        printf("Not found %d\n", delete_target);
    }

    return 0;
}
